core/src/main/java/com/google/instrumentation/tags/TagKey.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Copyright 2017, Google Inc.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.instrumentation.tags;
+
+import static com.google.instrumentation.tags.TagKey.TagType.TAG_BOOL;
+import static com.google.instrumentation.tags.TagKey.TagType.TAG_INT;
+import static com.google.instrumentation.tags.TagKey.TagType.TAG_STRING;
+
+import com.google.auto.value.AutoValue;
+import com.google.instrumentation.internal.StringUtil;
+import javax.annotation.concurrent.Immutable;
+
+/**
+ * A key to a value stored in a {@link TagSet}.
+ *
+ * @param <TagValueT> The type of value that can be paired with this {@code TagKey}. {@code TagKey}s
+ *     can only be instantiated with types {@code String}, {@code Long}, and {@code Boolean}.
+ */
+@Immutable
+@AutoValue
+public abstract class TagKey<TagValueT> {
+  /** The maximum length for a tag key name. */
+  public static final int MAX_LENGTH = StringUtil.MAX_LENGTH;
+
+  enum TagType {
+    TAG_STRING,
+    TAG_INT,
+    TAG_BOOL
+  }
+
+  TagKey() {}
+
+  /**
+   * Constructs a {@code TagKey<String>} from the given string. The string will be sanitized such
+   * that:
+   *
+   * <ol>
+   *   <li>length is restricted to {@link #MAX_LENGTH}, strings longer than that will be truncated.
```

Comment:
We should discuss what policies we want here. I think it would be better to raise an exception, rather than silently truncate or replace characters.